Tests/Unit/Consumer/AsyncQueueConsumerTest.php

```diff
@@ -0,0 +1,146 @@
+use Smartbox\Integration\FrameworkBundle\Tools\Helper\SmartesbHelper;
+
+/**
+ * Class AsyncQueueConsumerTest.
+ *
+ * @group async-queue-consumer
+ */
+class AsyncQueueConsumerTest extends TestCase
+{
+    use ConsumerMockFactory;
+
+    /**
+     * Assert that the consume function passes the correct objects to the driver.
+     */
+    public function testConsume()
+    {
+        $endpoint = $this->createMock(EndpointInterface::class);
+        $endpoint->expects($this->any())
+            ->method('getOptions')
+            ->willReturn(
+                [
+                    QueueProtocol::OPTION_QUEUE_DRIVER => 'ya-know-what?',
+                    'prefix' => 'doughnuts-',
+                    'queue' => 'should-be-plain',
+                ]
+            );
+
+        $driver = $this->createMock(AsyncQueueDriverInterface::class);
+        $driver->expects($this->once())
+            ->method('consume')
+            ->with(
+                $this->isType('string'),
+                $this->equalTo('doughnuts-should-be-plain'),
+                $this->isInstanceOf(\Closure::class)
+            );
+
+        // Consumer is extended with an anon class to fake the consumption of a message
+        $consumer = new class() extends AsyncQueueConsumer {
+            public function waitNoBlock(EndpointInterface $endpoint)
+            {
+                --$this->expirationCount;
+            }
+        };
+        $consumer->setSmartesbHelper($this->getHelper($driver));
+        $consumer->setExpirationCount(1);
+
+        $consumer->consume($endpoint, function () {});
+    }
+
+    /**
+     * Callback should call the exception handler when it fails to deserialize a message. By default, the
+     * ReThrowExceptionHandler is used which simply throws the exception again.
+     */
+    public function testUsesExceptionHandlerOnSerializationErrors()
+    {
+        $serializer = $this->createMock(SerializerInterface::class);
+        $serializer->expects($this->once())
+            ->method('deserialize')
+            ->willThrowException(new \RuntimeException('I cöuld nót dese�rialize that JSON strin��������'));
+
+        $driver = $this->createMock(AsyncQueueDriverInterface::class);
+        $driver->expects($this->once())
+            ->method('getFormat')
+            ->willReturn('json');
+
+        $messageHeaders = new AMQPTable([
+            'ttl' => 86400,
+            'expiration' => 86400000,
+            'expires' => 1584634937000,
+            'destination' => 'api-test',
+            'priority' => 4,
+        ]);
+
+        $message = $this->getMockBuilder(AMQPMessage::class)
+            ->setMethods(['setBody'])
+            ->getMock();
+        $message->method('setBody')->with('an amqp message');
+        $message->set('application_headers', $messageHeaders);
+        $message->delivery_info = ['delivery_tag' => 1];
+
+        $consumer = $this->getConsumer($this, AsyncQueueConsumer::class, $message, 1);
```

Comment:
When doing a unit testing, we shouldn't mock the class itself we are going to mock, otherwise the test diverts from the code is in the real class. Which is the reason on mocking the class itself?

Reply:
The callback is triggered deep in the driver, so unless this test uses the real driver (and the broker itself), there's no way to trigger it and get the message to perform assertions on it.

Injecting all the needed dependencies (endpoint and driver) mocked would create a lot of boilerplate code that can be simplified easily by mocking the tested function and overriding the parts we need.

`asyncConsume` in AsyncQueueConsumer would be extremely tricky to test it reliably:

```
        $queueName = $this->getQueueName($endpoint);
        $this->getQueueDriver($endpoint)->consume($this->getName(), $queueName, $callback);
```

You need an endpoint that will return the driver, that implements the consume function (which needs to call the callback in it)... It's tricky. That's why I think an easier approach is a mock.

Reply:
I'll admit I took the lazy approach, the correct way to test it would be with a really smart driver mock that would store the callback, trigger it when needed, and pass the correct information.

But I don't think we have enough time to rethink these tests. I'd say I took this approach because the functions originally were public and when we closed them I needed a way to keep the test more or less the same without changing much.

Reply:
Also (last comment, promise), given that only `consume()` is public, the test would be forced to go through the `process()` function which trigger a whole new level of dependencies, outside the scope of these tests. The amount of boilerplate code it would need would not only be massive, but prone to errors as well. Mocks are an easy way to bypass all of this.